File: frontend/src/lib/logger.ts
/**
 * Logger-Modul für strukturiertes Frontend-Logging
 * - DEV: Detaillierte Debug-Logs
 * - PROD: Nur Warnings und Errors
 *
 * Nutzung:
 *   import { logger } from "@/lib/logger";
 *   logger.debug("🔍 Debug info", { data });
 *   logger.info("✅ Success");
 *   logger.warn("⚠️ Warning");
 *   logger.error("❌ Error");
 */

type LogLevel = "DEBUG" | "INFO" | "WARN" | "ERROR";

class Logger {
  private isDev: boolean;
  private prefix: string;

  constructor() {
    // Check if we're in development
    this.isDev = process.env.NODE_ENV === "development";
    this.prefix = "🔧";
  }

  /**
   * Set a custom prefix for logs (e.g., "HTTP", "FORM", "INVOICE")
   */
  setPrefix(prefix: string): void {
    this.prefix = prefix;
  }

  /**
   * Format log output
   */
  private format(level: LogLevel, message: string, data?: unknown): string {
    const timestamp = new Date().toISOString().split("T")[1].slice(0, 8); // HH:MM:SS
    const levelColor = this.getLevelColor(level);
    const formatted = `[${timestamp}] ${levelColor} [${this.prefix}] ${message}`;
    return data ? `${formatted} ${JSON.stringify(data)}` : formatted;
  }

  /**
   * Get emoji representation for log level
   */
  private getLevelColor(level: LogLevel): string {
    const colors: Record<LogLevel, string> = {
      DEBUG: "🔍",
      INFO: "ℹ️",
      WARN: "⚠️",
      ERROR: "❌",
    };
    return colors[level];
  }

  /**
   * Debug log - only in development
   */
  debug(message: string, data?: unknown): void {
    if (!this.isDev) return;
    console.log(this.format("DEBUG", message, data));
  }

  /**
   * Info log - shown in dev and prod
   */
  info(message: string, data?: unknown): void {
    console.log(this.format("INFO", message, data));
  }

  /**
   * Warning log
   */
  warn(message: string, data?: unknown): void {
    console.warn(this.format("WARN", message, data));
  }

  /**
   * Error log
   */
  error(message: string, data?: unknown): void {
    console.error(this.format("ERROR", message, data));
  }

  /**
   * Create a scoped logger with a specific prefix
   */
  createScoped(prefix: string): ScopedLogger {
    return new ScopedLogger(prefix, this.isDev);
  }
}

/**
 * Scoped logger for specific modules/features
 */
class ScopedLogger {
  constructor(
    private prefix: string,
    private isDev: boolean
  ) {}

  private format(level: LogLevel, message: string, data?: unknown): string {
    const timestamp = new Date().toISOString().split("T")[1].slice(0, 8);
    const levelColor = this.getLevelColor(level);
    const formatted = `[${timestamp}] ${levelColor} [${this.prefix}] ${message}`;
    return data ? `${formatted} ${JSON.stringify(data)}` : formatted;
  }

  private getLevelColor(level: LogLevel): string {
    const colors: Record<LogLevel, string> = {
      DEBUG: "🔍",
      INFO: "ℹ️",
      WARN: "⚠️",
      ERROR: "❌",
    };
    return colors[level];
  }

  debug(message: string, data?: unknown): void {
    if (!this.isDev) return;
    console.log(this.format("DEBUG", message, data));
  }

  info(message: string, data?: unknown): void {
    console.log(this.format("INFO", message, data));
  }

  warn(message: string, data?: unknown): void {
    console.warn(this.format("WARN", message, data));
  }

  error(message: string, data?: unknown): void {
    console.error(this.format("ERROR", message, data));
  }
}

// Export global logger instance
export const logger = new Logger();

// Export for scoped usage
export { ScopedLogger };

File: backend/utils/logger.py
"""
Logger-Modul für strukturiertes Logging.
- DEV: Detaillierte Debug-Logs (DEBUG Level)
- PROD: Nur wichtige Info/Warnings (INFO Level)

Nutzung:
    from utils.logger import logger

    logger.debug("🔍 Debugging info")
    logger.info("✅ Success info")
    logger.warning("⚠️ Warning")
    logger.error("❌ Error")
"""

import logging
import os
from enum import Enum


class Environment(str, Enum):
    """Application environment enum."""

    DEV = "development"
    PROD = "production"


def get_environment() -> Environment:
    """Get the current environment from NODE_ENV or default to development."""
    env_str = os.getenv("NODE_ENV", "development").lower()
    return Environment.DEV if env_str in ["dev", "development"] else Environment.PROD


def setup_logger(name: str = "billino") -> logging.Logger:
    """
    Configure and return logger with environment-aware settings.

    Args:
        name: Logger name

    Returns:
        Configured logger instance
    """
    env = get_environment()

    # Determine log level based on environment
    if env == Environment.DEV:
        log_level = os.getenv("LOG_LEVEL", "DEBUG")
    else:
        log_level = os.getenv("LOG_LEVEL", "INFO")

    # Create logger
    logger = logging.getLogger(name)

    # Only configure if not already configured
    if not logger.handlers:
        logger.setLevel(log_level)

        # Create formatter with helpful prefixes
        formatter = logging.Formatter("[%(levelname)s] [%(name)s] %(message)s")

        # Console handler
        console_handler = logging.StreamHandler()
        console_handler.setFormatter(formatter)
        logger.addHandler(console_handler)

        # Log startup info
        logger.info(
            f"🚀 Logger initialized - Environment: {env.value} - Level: {log_level}"
        )

    return logger


# Create global logger instance
logger = setup_logger()
